libevmasm/Assembly.cpp

```diff
@@ -74,6 +75,130 @@ unsigned Assembly::codeSize(unsigned subTagSize) const
 	}
 }
 
+void Assembly::addAssemblyItemsFromJSON(Json::Value const& _code)
+{
+	solAssert(m_items.empty());
+	solAssert(_code.isArray());
+	for (auto const& jsonItem: _code)
+		m_items.emplace_back(createAssemblyItemFromJSON(jsonItem));
+
+	for (auto current = m_items.begin(); current != m_items.end(); ++current)
+	{
+		// During the assembly json export a `JUMPDEST` is always generated after a `tag`.
+		// So we just ignore exactly these `JUMPDEST`'s.
+		auto const nextItem = next(current);
+		if (nextItem != m_items.end() &&
+			current->type() == AssemblyItemType::Tag &&
+			nextItem->type() == AssemblyItemType::Operation &&
+			nextItem->instruction() == Instruction::JUMPDEST
+		)
+			m_items.erase(nextItem);
+	}
+
+	for (auto const& item: m_items)
+		if (item.type() == AssemblyItemType::Operation && item.instruction() == Instruction::JUMPDEST)
+			throw langutil::Error(
+				1285_error,
+				langutil::Error::Type::FatalError,
+				"JUMPDEST instruction found that was not followed by tag."
+			);
+}
+
+AssemblyItem Assembly::createAssemblyItemFromJSON(Json::Value const& _json)
+{
+	solAssert(isOfType<string>(_json["name"]));
+	solAssert(isOfType<int>(_json["begin"]));
+	solAssert(isOfType<int>(_json["end"]));
+	solAssert(isOfType<int>(_json["source"]));
+	solAssert(isOfTypeIfExists<string>(_json, "value"));
+	solAssert(isOfTypeIfExists<int>(_json, "modifierDepth"));
+	solAssert(isOfTypeIfExists<string>(_json, "jumpType"));
+
+	string name = getOrDefault<string>(_json["name"], "");
+	solAssert(!name.empty());
+
+	SourceLocation location;
+	location.start = get<int>(_json["begin"]);
+	location.end = get<int>(_json["end"]);
+	int srcIndex = get<int>(_json["source"]);
+	size_t modifierDepth = static_cast<size_t>(getOrDefault<int>(_json["modifierDepth"], 0));
+	string value = getOrDefault<string>(_json["value"], "");
+	string jumpType = getOrDefault<string>(_json["jumpType"], "");
+
+
+	auto updateUsedTags = [&](u256 const& data) {
+		m_usedTags = max(m_usedTags, static_cast<unsigned>(data) + 1);
+		return data;
+	};
+
+	auto immutableHash = [&](string const& _immutableName) -> h256 {
+		h256 hash(util::keccak256(_immutableName));
+		m_immutables[hash] = _immutableName;
```

Comment:
We generally don't expect to ever overwrite an existing hash with a different value, right?
```suggestion
		solAssert(m_immutables.count(hash) == 0 || m_immutables[hash] == _immutableName);
		m_immutables[hash] = _immutableName;
```

Same for library hash below.

Reply:
True